homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,432 @@
+
+async def async_setup(hass, config):
+    """Set up the scheduler."""
+    if DOMAIN not in config:
+        return True
+
+    scheduler = hass.data[DOMAIN] = Scheduler(hass)
+    await scheduler.async_load()
+
+    hass.components.websocket_api.async_register_command(
+        WS_TYPE_SCHEDULER_CREATE_SCHEDULE,
+        websocket_handle_create,
+        SCHEMA_WEBSOCKET_CREATE_SCHEDULE,
+    )
+    hass.components.websocket_api.async_register_command(
+        WS_TYPE_SCHEDULER_DELETE_SCHEDULE,
+        websocket_handle_delete,
+        SCHEMA_WEBSOCKET_DELETE_SCHEDULE,
+    )
+    hass.components.websocket_api.async_register_command(
+        WS_TYPE_SCHEDULER_SCHEDULES,
+        websocket_handle_schedules,
+        SCHEMA_WEBSOCKET_SCHEDULES,
+    )
+    hass.components.websocket_api.async_register_command(
+        WS_TYPE_SCHEDULER_UPDATE_SCHEDULE,
+        websocket_handle_update,
+        SCHEMA_WEBSOCKET_UPDATE_SCHEDULE,
+    )
+
+    return True
+
+
+class ScheduleInstance:
+    """A class that defines a scene schedule."""
+
+    def __init__(
+        self, hass: "HomeAssistant", entity_id: str, activation_context: Context,
+    ) -> None:
+        """Initialize."""
+        self._activation_context = activation_context
+        self._async_state_listener: Optional[Callable] = None
+        self._entity_states: Dict[str, "State"] = {}
+        self._hass: "HomeAssistant" = hass
+        self.entity_id: str = entity_id
+        self.instance_id = uuid4().hex
+
+    @callback
+    def async_trigger_scene(self) -> None:
+        """Trigger the schedule's scene."""
+        self._hass.async_create_task(
+            self._hass.services.async_call(
+                "scene",
+                "turn_on",
+                service_data={"entity_id": self.entity_id},
+                context=self._activation_context,
+            )
+        )
+
+        @callback
+        def store_entity_if_in_context(
+            entity_id: str, old_state: "State", new_state: "State"
+        ) -> None:
+            """Save prior states of an entity if it was triggered by this schedule."""
+            if new_state.context == self._activation_context:
+                self._entity_states[entity_id] = old_state
+
+        self._async_state_listener = async_track_state_change(
+            self._hass, MATCH_ALL, store_entity_if_in_context
+        )
+
+        _LOGGER.info("Scheduler activated scene: %s", self.entity_id)
+
+    @callback
+    def async_undo_scene(self) -> None:
+        """Restore the entities touched by the schedule."""
+        if not self._entity_states:
+            return
+
+        for entity_id, state in self._entity_states.items():
+            self._hass.states.async_set(
+                entity_id,
+                state.state,
+                attributes=state.attributes,
+                context=self._activation_context,
+            )
+
+        _LOGGER.info("Scheduler deactivated scene: %s", self.entity_id)
+
+
+class Scheduler:
```

Comment:
This class is conflating concerns. Each schedule should be its own class which can have an active instance. 

`Scheduler -> Schedule -> ActiveSchedule`.